File: src/simplesave/engine/risk.py
"""Weighted risk score for a mix (reference ``mixRisk``)."""

from __future__ import annotations

from simplesave.engine.core import js_round, num
from simplesave.engine.tuning import infer_route_kind, route_change_period
from simplesave.engine.types import IndexType, RiskResult, RiskRule, Route, RouteKind


def default_risk_rules() -> list[RiskRule]:
    """Reference ``defaultRiskRules``."""
    return [
        RiskRule(RouteKind.PRIME.value, 1, 12, "לא", "נמוך", 1),
        RiskRule(RouteKind.VARIABLE.value, 1, 59, "לא", "בינוני", 2),
        RiskRule(RouteKind.VARIABLE.value, 1, 59, "כן", "בינוני", 3),
        RiskRule(RouteKind.VARIABLE.value, 60, 360, "לא", "גבוה", 3),
        RiskRule(RouteKind.VARIABLE.value, 60, 360, "כן", "גבוה", 4),
        RiskRule(RouteKind.FIXED.value, 48, 360, "לא", "גבוה", 3),
        RiskRule(RouteKind.FIXED.value, 48, 360, "כן", "גבוה", 4),
    ]


def risk_rule_for_route(route: Route, rules: list[RiskRule]) -> RiskRule:
    """Reference ``riskRuleForRoute``: match a route to its risk rule."""
    months = route_change_period(route)
    indexed = "כן" if route.index_type == IndexType.CPI else "לא"
    kind = infer_route_kind(route).value

    for rule in rules:
        if (
            (rule.route_kind == "all" or rule.route_kind == kind)
            and months >= num(rule.from_months)
            and months <= num(rule.to_months)
            and (rule.indexed == "הכול" or rule.indexed == indexed)
        ):
            return rule
    for rule in rules:
        if (rule.route_kind == "all" or rule.route_kind == kind) and (
            rule.indexed == "הכול" or rule.indexed == indexed
        ):
            return rule
    return RiskRule("all", 0, 0, "הכול", "נמוך", 1)


def mix_risk(routes: list[Route], rules: list[RiskRule] | None = None) -> RiskResult:
    """Reference ``mixRisk``: share- (or amount-) weighted risk score."""
    if rules is None:
        rules = default_risk_rules()

    use_shares = sum(num(rt.share_pct) for rt in routes) > 0

    def weight(rt: Route) -> float:
        return num(rt.share_pct) if use_shares else num(rt.amount)

    total = sum(weight(rt) for rt in routes)
    if total <= 0:
        return RiskResult(score=0.0, level=0, label="—")

    score = sum(weight(rt) * num(risk_rule_for_route(rt, rules).risk) for rt in routes) / total
    level = min(5, max(1, js_round(score)))
    if score < 1.75:
        label = "נמוכה"
    elif score < 2.75:
        label = "בינונית"
    elif score < 3.75:
        label = "גבוהה"
    else:
        label = "גבוהה מאוד"
    return RiskResult(score=score, level=level, label=label)


__all__ = ["default_risk_rules", "mix_risk", "risk_rule_for_route"]

File: lib/engine/risk.ts
// Weighted risk score for a mortgage mix.

import { jsRound, num } from "./core";
import { inferRouteKind, routeChangePeriod } from "./tuning";
import type { RiskResult, RiskRule, Route } from "./types";

export function defaultRiskRules(): RiskRule[] {
  return [
    { routeKind: "prime", fromMonths: 1, toMonths: 12, indexed: "לא", exitPenalty: "נמוך", risk: 1 },
    { routeKind: "variable", fromMonths: 1, toMonths: 59, indexed: "לא", exitPenalty: "בינוני", risk: 2 },
    { routeKind: "variable", fromMonths: 1, toMonths: 59, indexed: "כן", exitPenalty: "בינוני", risk: 3 },
    { routeKind: "variable", fromMonths: 60, toMonths: 360, indexed: "לא", exitPenalty: "גבוה", risk: 3 },
    { routeKind: "variable", fromMonths: 60, toMonths: 360, indexed: "כן", exitPenalty: "גבוה", risk: 4 },
    { routeKind: "fixed", fromMonths: 48, toMonths: 360, indexed: "לא", exitPenalty: "גבוה", risk: 3 },
    { routeKind: "fixed", fromMonths: 48, toMonths: 360, indexed: "כן", exitPenalty: "גבוה", risk: 4 },
  ];
}

export function riskRuleForRoute(route: Route, rules: RiskRule[]): RiskRule {
  const months = routeChangePeriod(route);
  const indexed = route.indexType === "מדד" ? "כן" : "לא";
  const kind = inferRouteKind(route);

  for (const rule of rules) {
    if (
      (rule.routeKind === "all" || rule.routeKind === kind) &&
      months >= num(rule.fromMonths) &&
      months <= num(rule.toMonths) &&
      (rule.indexed === "הכול" || rule.indexed === indexed)
    ) {
      return rule;
    }
  }
  for (const rule of rules) {
    if (
      (rule.routeKind === "all" || rule.routeKind === kind) &&
      (rule.indexed === "הכול" || rule.indexed === indexed)
    ) {
      return rule;
    }
  }
  return { routeKind: "all", fromMonths: 0, toMonths: 0, indexed: "הכול", exitPenalty: "נמוך", risk: 1 };
}

export function mixRisk(routes: Route[], rules: RiskRule[] = defaultRiskRules()): RiskResult {
  const useShares = routes.reduce((s, rt) => s + num(rt.sharePct), 0) > 0;
  const weight = (rt: Route) => (useShares ? num(rt.sharePct) : num(rt.amount));
  const total = routes.reduce((s, rt) => s + weight(rt), 0);
  if (total <= 0) return { score: 0, level: 0, label: "—" };

  const score = routes.reduce((s, rt) => s + weight(rt) * num(riskRuleForRoute(rt, rules).risk), 0) / total;
  const level = Math.min(5, Math.max(1, jsRound(score)));
  const label = score < 1.75 ? "נמוכה" : score < 2.75 ? "בינונית" : score < 3.75 ? "גבוהה" : "גבוהה מאוד";
  return { score, level, label };
}
